### DISCUSSION TITLE
Ordering foreign table rows

### DISCUSSION THREAD

**Discussion by Polyterative** [supabase/supabase#1690] [Questions] [answered]
I have this in the docs
```js
READ FOREIGN TABLES
let { data: module_outs, error } = await supabase
  .from('module_outs')
  .select(`
    some_column,
    other_table (
      foreign_key
    )
  `)
  ```
  
  and it's fine. 
  I'm trying to order the results of the foreign table result. 
  
  Now I'm here
  I want to order the result of the pink part by `id`
  
![image](https://user-images.githubusercontent.com/16295552/119261841-847cc980-bbd9-11eb-9736-60ae64963716.png)

can't find any way to do it, any help please?

**Comment by soedirgo** [comment, marked as answer]
Hey @Polyterative, you can order nested foreign tables, like so:
```js
const { data, error } = await supabase
    .from('module_outs')
    .select(`
        some_column,
        other_table (
            yet_another_table ( id )
        )
    `)
    .order('id', { foreignTable: 'other_table.yet_another_table' })
```

**Reply by Polyterative** [reply]
didn't know about that, will try, thanks!

**Reply by donPuerto** [reply]
@soedirgo What this id mean? Is this the primary key ID of the parent table?
![image](https://user-images.githubusercontent.com/9329005/129300018-2fcd9c3a-3b67-4620-9eb4-7d7e692166b3.png)

How to do ascending/descending on the foreign table?

**Reply by soedirgo** [reply]
It's the column of the nested foreign table that you want to order by. You can do ascending/descending by setting `ascending` to true/false like the docs here: https://supabase.io/docs/reference/javascript/order

**Reply by richcorbs** [reply]
Does this still work as described above?

I'm using the exact same syntax and it always returns the foreign table in the order they were created.

Maybe I'm misunderstanding.

I'm trying to order by `module_outs.id` and then by `yet_another_table.id`. The sort by `module_outs.id` works just right but then it's like the second order is ignored.

```
({ data, error } = await supabase
  .from('module_outs')
  .select('*,other_table(*,yet_another_table(*))')
  .order('id', { ascending: false })
  .order('id', { foreignTable: 'other_table.yet_another_table', ascending: true })
```
Even if I remove the first order the second doesn't seem to make a difference.

**Reply by soedirgo** [reply]
@richcorbs Yeah, it should still work:
```sql
create table table1 (
  id int4 generated by default as identity primary key
);

insert into table1(id) values (default);

create table table2 (
  id int4 generated by default as identity primary key,
  table1_id int4 references table1(id)
);

insert into table2(table1_id) values (1), (1);

create table table3 (
  id int4 generated by default as identity primary key,
  table2_id int4 references table2(id)
);

insert into table3(table2_id) values (1), (1);
```
```js
const { createClient } = require('@supabase/supabase-js')
const supabase = createClient(..., ...);
supabase
    .from('table1')
    .select('*, table2(*, table3(*))')
    .order('id', { foreignTable: 'table2', ascending: false })
    .order('id', { foreignTable: 'table2.table3', ascending: false })
    .then(({ data }) => console.log(JSON.stringify(data, null, 2)))
```
```json
[
  {
    "id": 1,
    "table2": [
      {
        "id": 2,
        "table1_id": 1,
        "table3": []
      },
      {
        "id": 1,
        "table1_id": 1,
        "table3": [
          {
            "id": 2,
            "table2_id": 1
          },
          {
            "id": 1,
            "table2_id": 1
          }
        ]
      }
    ]
  }
]
```

**Reply by richcorbs** [reply]
Thank you for the response. @soedirgo!

I think I found the difference. My schema is as follows. I think I'm trying to sort in a way that I can't.

```
create table table1 (
  id int4 generated by default as identity primary key
);

create table table3 (
  id int4 generated by default as identity primary key,
  number int4
);

create table table2 (
  id int4 generated by default as identity primary key,  
  table1_id int4 references table1(id),
  table3_id int4 references table3(id)
);

insert into table1(id) values (default);

insert into table3(number) values (0), (1);

insert into table2(table1_id, table3_id) values (1, 1), (1, 2);

// JSON
[
  {
    "id": 1,
    "table2": [
      {
        "id": 2,
        "table1_id": 1,
        "table3_id": 2,
        "table3": {
          "id": 2,
          "number": 1
        }
      },
      {
        "id": 1,
        "table1_id": 1,
        "table3_id": 1,
        "table3": {
          "id": 1,
          "number": 0
        }
      }
    ]
  }
]
```
I was hoping that I could sort the table2 entries by `table3.number`. From my current understanding that's not possible.

If I had multiple table3 entries per table2 record I could sort them within the table2 record but not across table2 records like I'm trying.

Thanks for reading!

**Reply by soedirgo** [reply]
For folks having the same issue as above, this is discussed here: https://github.com/supabase/supabase/discussions/2990
